torch/distributed/rpc/api.py

```diff
@@ -83,122 +83,135 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
-class WaitAllWorkersStates(object):
+class AllGatherStates(object):
     def __init__(self):
-        # Each `intent_worker_names` is an empty set at beginning.
-        # It's only used by leader worker. Leader worker is user-specified or
-        # elected as the first worker in a sorted worker name list.
-        # Whenever there is a worker showing shutdown intention to the leader, by
-        # calling `_wait_all_workers()`, the leader adds this worker's name to the set.
-        # The leader also adds itself's name to the set on calling
-        # `_wait_all_workers()`. We need this because, we confine `_wait_all_workers()`
-        # to be called only once, by examing if leader's name has been added to the set.
-        self.intent_worker_names = set()
-        # Once `intent_worker_names == _ALL_WORKER_NAMES`,
-        # we flip `_SHUTDOWN_PROCEED_SIGNAL` on the leader, and leader will send RPCs
-        # to follower workers to flip their `_SHUTDOWN_PROCEED_SIGNAL`s.
-        self.proceed_signal = threading.Event()
-
-
-# States used by `def _wait_all_workers()`.
+        # Each `gathered_objects` is an empty dict at beginning.
+        # It's only used by the leader worker. The leader worker is elected as
+        # the first worker in a sorted worker name list.
+        # Whenever there is a worker entering _all_gather(), it runs
+        # `_gather_to_leader` on the leader to add its own name and data obj to
+        # this dict.
+        # The leader also adds itself's name to the dict on calling
+        # `_all_gather()`.
+        self.gathered_objects = {}
+        # Once `set(gathered_objects.keys()) == _ALL_WORKER_NAMES`, the leader
+        # will broadcast the gathered dict to all follower workers and unblock
+        # them.
+        self.results_future = torch.futures.Future()
```

Comment:
Nit: I think we could stick to having a `threading.Event` here rather than a future. A future is basically just an Event plus a value. And we already have a separate field for storing the value: `gathered_object`. So having both that field and this future may make it unclear where the value really resides and may lead to the value being copied twice. In fact, on the leader, we do copy the value twice, whereas on the other workers we don't use `gathered_objects` at all. It may be simpler to always keep the value in `gathered_objects` only and use an event here just for signaling purposes.

Reply:
Good point! will update.